src/python/pants/backend/jvm/tasks/junit_run.py

```diff
@@ -506,57 +512,106 @@ def _validate_target(self, target):
       msg = 'JUnitTests target must include a non-empty set of sources.'
       raise TargetDefinitionException(target, msg)
 
+  @staticmethod
+  def _vts_for_partition(invalidation_check):
+    return VersionedTargetSet.from_versioned_targets(invalidation_check.all_vts)
+
+  def check_artifact_cache_for(self, invalidation_check):
+    # We generate artifacts, namely coverage reports, that cover the full target set.
+    return [self._vts_for_partition(invalidation_check)]
+
   def _execute(self, all_targets):
     # NB: We only run tests within junit_tests targets, but if coverage options are
     # specified, we want to instrument and report on all the original targets, not
     # just the test targets.
-
-    test_registry = self._collect_test_targets(self._get_test_targets())
-    if test_registry.empty:
-      return
-
-    with self._isolation(all_targets) as (output_dir, do_report, coverage):
-      try:
-        self._run_tests(test_registry, output_dir, coverage)
-        do_report(exc=None)
-      except TaskError as e:
-        do_report(exc=e)
-        raise
+    partition = all_targets if self.get_options().coverage else self._get_test_targets()
+
+    with self.invalidated(targets=partition,
+                          # Re-run tests when the code they test (and depend on) changes.
+                          invalidate_dependents=True) as invalidation_check:
+
+      is_test_target = self._test_target_filter()
+      invalid_test_tgts = [invalid_tgt
+                           for vts in invalidation_check.invalid_vts
+                           for invalid_tgt in vts.targets if is_test_target(invalid_tgt)]
+
+      test_registry = self._collect_test_targets(invalid_test_tgts)
+
+      # Processing proceeds through:
+      # 1.) output -> output_dir
+      # 2.) [iff all == invalid] output_dir -> cache: We do this manually for now.
+      # 3.) [iff invalid == 0 and all > 0] cache -> workdir: Done transparently by `invalidated`.
+      # 4.) [iff user-specified final locations] workdir -> final-locations: We perform this step
+      #     as an unconditional post-process in `_isolation`.
+      with self._isolation(all_targets) as (output_dir, reports, coverage):
+        if not test_registry.empty:
+          try:
+            # 1.) Write all results that will be potentially cached to output_dir.
+            self._run_tests(test_registry, output_dir, coverage)
+            reports.generate()
+
+            cache_vts = self._vts_for_partition(invalidation_check)
+            if invalidation_check.all_vts == invalidation_check.invalid_vts:
+              # 2.) The full partition was invalid, cache results.
+              if self.artifact_cache_writes_enabled():
```

Comment:
Done - although I kept the same structure as `tasks2/PytestRun`, so not a full extraction. The refactor to lift up generic 1..N partition caching will finish this off.